File: java/po/world_simulator/World.java
package po.world_simulator;

import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Random;
import java.util.function.Function;
import po.world_simulator.entities.Entity;
import po.world_simulator.entities.animals.*;
import po.world_simulator.entities.plants.*;

public class World {
    private PriorityQueue<Entity> entities;
    private PriorityQueue<Entity> nextTurnEntities;

    private Vector2 worldSize;
    private Boolean isHexMode;

    private po.world_simulator.Map map;
    private Renderer renderer;
    private InputManager inputManager;

    private static World instance;
    private Random random = new Random();

    public World(boolean hexMode) {
        worldSize = new Vector2(Config.MAP_SIZE_X, Config.MAP_SIZE_Y);
        this.isHexMode = hexMode;

        this.map = new po.world_simulator.Map(worldSize, hexMode);
        loadRenderer();
        this.inputManager = new InputManager();

        if (instance != null) {
            instance.dispose();
        }

        instance = this;

        entities = new PriorityQueue<>(java.util.Collections.reverseOrder());
        nextTurnEntities = new PriorityQueue<>(java.util.Collections.reverseOrder());

        createEntities();
    }

    public void dispose() {
        clearEntities();
        renderer.dispose();
        if(instance == this)
            instance = null;
    }

    public static Renderer getRenderer() {
        return instance.renderer;
    }

    public static po.world_simulator.Map getMap() {
        return instance.map;
    }

    public static InputManager getInputManager() {
        return instance.inputManager;
    }

    public static World getInstance() {
        return instance;
    }

    public static int getEntityCount() {
        return instance.entities.size();
    }

    private void loadRenderer(){
        if(renderer != null)
            renderer.dispose();
        renderer = isHexMode ? new RendererHex(worldSize) : new Renderer(worldSize);

        // bind buttons
        renderer.bind("Save", () -> World.getInputManager().pushInput((int) Config.SAVE));
        renderer.bind("Load", () -> World.getInputManager().pushInput((int) Config.LOAD));
    }

    public boolean performTurn() {
        do {
            inputManager.readNextInput();

            if (inputManager.isQuitPressed())
                return false;

            if (inputManager.getLastInput() == Config.SAVE) {
                save();
                inputManager.clearLastInput();
                // skip turn after saving
                return true;
            } else if (inputManager.getLastInput() == Config.LOAD) {
                load();
                inputManager.clearLastInput();
                // skip turn after loading
                return true;
            }

            // redraw info window to show pressed input
            renderer.drawInfoWindow();
            renderer.renderInfoWindow();
        }
        while (!inputManager.isNewTurnPressed());

        renderer.clearLog();

        while (!entities.isEmpty()) {
            Entity entity = entities.poll();

            if (entity.isAlive())
                entity.action();

            queueToNext(entity);
        }

        nextQueue();
        return true;
    }

    public void drawWorld() {
        renderer.drawMap(worldSize);

        while (!entities.isEmpty()) {
            Entity entity = entities.poll();

            if (entity.isAlive()) entity.draw();

            queueToNext(entity);
        }

        renderer.renderMap();

        nextQueue();

        renderer.drawInfoWindow();
        renderer.renderInfoWindow();
    }

    public void forceDrawAll(){
        // force push all entities that were already added
        nextQueue();

        // draw world with all entities added
        drawWorld();
    }

    public void save() {
        SaveParser parser = new SaveParser();

        parser.createFile(worldSize);

        parser.addEntry("world_size", DataFormat.TYPE_VECTOR2, worldSize.toString());
        parser.addEntry("hex_mode", DataFormat.TYPE_BOOLEAN, isHexMode.toString());

        parser.startCategory("entities");

        while (!entities.isEmpty()) {
            Entity entity = entities.poll();

            if (entity.isAlive()) parser.addStringEntry(entity.saveAsString(parser));

            queueToNext(entity);
        }

        nextQueue();

        parser.endCategory();

        parser.closeFile();

        renderer.clearLog();
        renderer.addToLog("game saved!");
    }

    public void load() {
        renderer.clearLog();

        SaveParser parser = new SaveParser();

        if (!parser.loadFile(worldSize)) {
            renderer.addToLog("no save file found!");
            return;
        }

        worldSize = (Vector2)parser.loadEntry("world_size");
        boolean hexMode = (boolean)parser.loadEntry("hex_mode");

        // clear previous world
        map = new po.world_simulator.Map(worldSize, isHexMode);
        // reload renderer if save contains different grid mode
        if(hexMode != isHexMode){
            isHexMode = hexMode;
            loadRenderer();
        }
        clearEntities();

        // load entities
        if (!parser.jumpToCategory("entities")) {
            renderer.addToLog("incorrect save file format");
            return;
        }

        Map<String, String> entityData = new HashMap<>();

        while (parser.loadEntryMultiline(entityData)) {
            loadEntity(entityData);
            entityData.clear();
        }

        nextQueue();
        renderer.addToLog("game loaded!");
    }

    public void addNewEntity(Entity entity) {
        if (map.isTileOccupied(entity.getPosition())) {
            // convert entity into kebap
            entity.dispose();
        } else {
            // add entity to world
            map.placeEntityAt(entity.getPosition(), entity);
            queueToNext(entity);
        }
    }


    private void queueToNext(Entity entity) {
        if (entity.isAlive())
            nextTurnEntities.add(entity);
        else
            entity.dispose();
    }

    private void nextQueue() {
        entities.addAll(nextTurnEntities);
        nextTurnEntities.clear();
    }

    private void createEntities() {
        int worldArea = worldSize.x * worldSize.y;

        addNewEntity(new Human(worldSize.multiply(0.5f)));

        randomizeEntities(Grass::new, worldArea, Config.GRASS_AMOUNT);
        randomizeEntities(Milkweed::new, worldArea, Config.MILKWEED_AMOUNT);
        randomizeEntities(Guarana::new, worldArea, Config.GUARANA_AMOUNT);
        randomizeEntities(Wolfberries::new, worldArea, Config.WOLFBERRIES_AMOUNT);
        randomizeEntities(SosnowskiHogweed::new, worldArea, Config.SOSNOWSKI_HOGWEED_AMOUNT);
        randomizeEntities(Wolf::new, worldArea, Config.WOLF_AMOUNT);
        randomizeEntities(Sheep::new, worldArea, Config.SHEEP_AMOUNT);
        randomizeEntities(Fox::new, worldArea, Config.FOX_AMOUNT);
        randomizeEntities(Turtle::new, worldArea, Config.TURTLE_AMOUNT);
        randomizeEntities(Antelope::new, worldArea, Config.ANTELOPE_AMOUNT);

        nextQueue();
    }

    private void randomizeEntities(Function<Vector2, Entity> constructor, int worldArea, int percent) {
        int amount = worldArea * percent / 100;
        for (int i = 0; i < amount; ++i) {
            addNewEntity(constructor.apply(new Vector2(random.nextInt(worldSize.x), random.nextInt(worldSize.y))));
        }
    }

    public void loadEntity(Map<String, String> entityData) {
        String type = entityData.get("type");

        String[] packages = {
            "po.world_simulator.entities.animals.",
            "po.world_simulator.entities.plants."
        };

        for (String pkg : packages) {
            try {
                Class<?> clazz = Class.forName(pkg + type);
                Entity entity = (Entity) clazz.getDeclaredConstructor(Map.class).newInstance(entityData);
                addNewEntity(entity);
                return;
            } catch (Exception ignored) {
            }
        }

        renderer.addToLog("UNKNOWN ENTITY TYPE: " + type);
    }


    private void clearEntities() {
        while (!entities.isEmpty()) {
            Entity entity = entities.poll();
            if (entity != null) entity.dispose();
        }
        while (!nextTurnEntities.isEmpty()) {
            Entity entity = nextTurnEntities.poll();
            if (entity != null) entity.dispose();
        }
    }
}

File: python/po/world_simulator/World.py
import heapq
import random
import importlib
import sys

from .Vector2 import Vector2
from .Config import Config
from .Map import Map as WorldMap
from .InputManager import InputManager
from .SaveParser import SaveParser
from .DataFormat import DataFormat

from po.world_simulator.entities.animals.Human import Human
from po.world_simulator.entities.plants.Grass import Grass
from po.world_simulator.entities.plants.Milkweed import Milkweed
from po.world_simulator.entities.plants.Guarana import Guarana
from po.world_simulator.entities.plants.Wolfberries import Wolfberries
from po.world_simulator.entities.plants.SosnowskiHogweed import SosnowskiHogweed
from po.world_simulator.entities.animals.Wolf import Wolf
from po.world_simulator.entities.animals.Sheep import Sheep
from po.world_simulator.entities.animals.Fox import Fox
from po.world_simulator.entities.animals.Turtle import Turtle
from po.world_simulator.entities.animals.Antelope import Antelope
from po.world_simulator.entities.animals.CyberSheep import CyberSheep

class World:
    instance = None

    def __init__(self, hexMode):
        self.worldSize = Vector2(Config.MAP_SIZE_X, Config.MAP_SIZE_Y)
        self.isHexMode = hexMode

        self.map = WorldMap(self.worldSize, hexMode)
        self.renderer = None
        self.loadRenderer()
        self.inputManager = InputManager()

        if World.instance is not None:
            World.instance.dispose()

        World.instance = self

        # Using list for PriorityQueue. Will use heapq module for ops.
        self.entities = []
        self.hogweed: list[SosnowskiHogweed] = []
        self.nextTurnEntities = []

        self.createEntities()


    def dispose(self):
        self.clearEntities()
       self.renderer.dispose()
        if World.instance == self:
            World.instance = None


    @staticmethod
    def getRenderer():
        return World.instance.renderer


    @staticmethod
    def getMap():
        return World.instance.map


    @staticmethod
    def getInputManager():
        return World.instance.inputManager


    @staticmethod
    def getInstance():
        return World.instance


    @staticmethod
    def getEntityCount():
        return len(World.instance.entities)


    @staticmethod
    def getHogweed() -> list[SosnowskiHogweed]:
        return World.instance.hogweed


    def loadRenderer(self):
        if self.renderer is not None:
            self.renderer.dispose()

        if self.isHexMode:
            from .RendererHex import RendererHex
            self.renderer = RendererHex(self.worldSize)
        else:
            from .Renderer import Renderer
            self.renderer = Renderer(self.worldSize)

        # bind buttons
        self.renderer.bind("Save", lambda: World.getInputManager().pushInput(Config.SAVE))
        self.renderer.bind("Load", lambda: World.getInputManager().pushInput(Config.LOAD))


    def performTurn(self):
        while True:
            self.inputManager.readNextInput()

            if self.inputManager.isQuitPressed():
                return False

            if self.inputManager.getLastInput() == Config.SAVE:
                self.save()
                self.inputManager.clearLastInput()
                return True
            elif self.inputManager.getLastInput() == Config.LOAD:
                self.load()
                self.inputManager.clearLastInput()
                # skip turn after loading
                return True

            # redraw info window to show pressed input
            self.renderer.drawInfoWindow()
            self.renderer.renderInfoWindow()

            if self.inputManager.isNewTurnPressed():
                break

        self.renderer.clearLog()

        while self.entities:
            entity = heapq.heappop(self.entities)

            if entity.isAlive():
                entity.action()

            self.queueToNext(entity)

        self.nextQueue()
        return True

    def drawWorld(self):
        self.renderer.drawMap(self.worldSize)

        while self.entities:
            entity = heapq.heappop(self.entities)

            if entity.isAlive() and self.renderer:
                entity.draw()

            self.queueToNext(entity)

        self.renderer.renderMap()

        self.nextQueue()

        self.renderer.drawInfoWindow()
        self.renderer.renderInfoWindow()


    def forceDrawAll(self):
        # force push all entities that were already added
        self.nextQueue()

        # draw world with all entities added
        self.drawWorld()


    def save(self):
        parser = SaveParser()

        parser.createFile(self.worldSize)

        parser.addEntry("world_size", DataFormat.TYPE_VECTOR2, str(self.worldSize))
        parser.addEntry("hex_mode", DataFormat.TYPE_BOOLEAN, str(self.isHexMode).lower())

        parser.startCategory("entities")

        while self.entities:
            entity = heapq.heappop(self.entities)

            if entity.isAlive():
                parser.addStringEntry(entity.saveAsString(parser))

            self.queueToNext(entity)

        self.nextQueue()

        parser.endCategory()

        parser.closeFile()

        self.renderer.clearLog()
        self.renderer.addToLog("game saved!")


    def load(self):
        self.renderer.clearLog()

        parser = SaveParser()

        if not parser.loadFile(self.worldSize):
            self.renderer.addToLog("no save file found!")
            return

        self.worldSize = parser.loadEntry("world_size")
        hexMode = parser.loadEntry("hex_mode")

        # clear previous world
        self.map = WorldMap(self.worldSize, self.isHexMode)
        # reload renderer if save contains different grid mode
        if hexMode != self.isHexMode:
            self.isHexMode = hexMode
            self.loadRenderer()

        self.clearEntities()

        # load entities
        if not parser.jumpToCategory("entities"):
            self.renderer.addToLog("incorrect save file format")
            return

        entityData = {}

        while parser.loadEntryMultiline(entityData):
            self.loadEntity(entityData)
            entityData.clear()

        self.nextQueue()
        iself.renderer.addToLog("game loaded!")


    def addNewEntity(self, entity):
        if self.map.isTileOccupied(entity.getPosition()):
            # convert entity into kebap
            entity.dispose()
        else:
            # add entity to world
            self.map.placeEntityAt(entity.getPosition(), entity)
            self.queueToNext(entity)


    def queueToNext(self, entity):
        if entity.isAlive():
            self.nextTurnEntities.append(entity)
        else:
            entity.dispose()


    def nextQueue(self):
        self.hogweed.clear()
        for entity in self.nextTurnEntities:
            heapq.heappush(self.entities, entity)
            if isinstance(entity, SosnowskiHogweed):
                self.hogweed.append(entity)
        self.nextTurnEntities.clear()


    def createEntities(self):
        worldArea = self.worldSize.x * self.worldSize.y

        self.addNewEntity(Human(self.worldSize.multiply(0.5)))

        self.randomizeEntities(Grass, worldArea, Config.GRASS_AMOUNT)
        self.randomizeEntities(Milkweed, worldArea, Config.MILKWEED_AMOUNT)
        self.randomizeEntities(Guarana, worldArea, Config.GUARANA_AMOUNT)
        self.randomizeEntities(Wolfberries, worldArea, Config.WOLFBERRIES_AMOUNT)
        self.randomizeEntities(SosnowskiHogweed, worldArea, Config.SOSNOWSKI_HOGWEED_AMOUNT)
        self.randomizeEntities(Wolf, worldArea, Config.WOLF_AMOUNT)
        self.randomizeEntities(Sheep, worldArea, Config.SHEEP_AMOUNT)
        self.randomizeEntities(Fox, worldArea, Config.FOX_AMOUNT)
        self.randomizeEntities(Turtle, worldArea, Config.TURTLE_AMOUNT)
        self.randomizeEntities(Antelope, worldArea, Config.ANTELOPE_AMOUNT)
        self.randomizeEntities(CyberSheep, worldArea, Config.CYBER_SHEEP_AMOUNT)

        self.nextQueue()


    def randomizeEntities(self, constructor, worldArea, percent):
        amount = int(worldArea * percent / 100)
        for _ in range(amount):
            self.addNewEntity(constructor(Vector2(random.randint(0, self.worldSize.x - 1), random.randint(0, self.worldSize.y - 1))))


    def loadEntity(self, entityData):
        type_str = entityData.get("type")

        packages = [
            "po.world_simulator.entities.animals",
            "po.world_simulator.entities.plants"
        ]

        for pkg in packages:
            try:
                module = importlib.import_module(f"{pkg}.{type_str}")
                clazz = getattr(module, type_str)
                
                entity = clazz(entityData)
                self.addNewEntity(entity)
                return
            except Exception:
                pass

        self.renderer.addToLog("UNKNOWN ENTITY TYPE: " + str(type_str))


    def clearEntities(self):
        while self.entities:
            entity = heapq.heappop(self.entities)
            if entity is not None:
                entity.dispose()

        while self.nextTurnEntities:
            entity = self.nextTurnEntities.pop(0)
            if entity is not None:
                entity.dispose()